Add file size columns to download status table

ANAMASGARD opened the pull request:
Fixes #249

## Summary
Adds file size information (total_MB, mean_MB, and rows) to the download status table, helping users understand disk usage for each geom.

## Problem
Currently, the download status table shows how many TSV files need to be downloaded, but doesn't show file sizes. Users can't see if they're downloading 1MB or 100MB.

**Test output:**
<img width="3072" height="1840" alt="Screenshot From 2025-10-29 10-50-36" src="https://github.com/user-attachments/assets/e3e75fb3-a6e3-467f-b361-3116e77872b9" />

Only added the test case in this first commit in the next commit will try to fix the issue !

ANAMASGARD committed:
Add test for download status table size columns

Test verifies animint.js generates total_MB and mean_MB columns
in download status table. Currently fails, demonstrating these
size columns are not present in the table.

Addresses #249

ANAMASGARD committed:
Now using grep() and expect_gt() for informative test failures

ANAMASGARD committed:
Added the download size info to status table

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -1008,6 +1011,37 @@ var animint = function (to_select, json_file) {
+    }
+    var total_MB = (total_bytes / 1048576).toFixed(2);
+    var mean_MB = chunk_count > 0 ? (total_bytes / chunk_count / 1048576).toFixed(2) : "0.00";
+    g_info.tr.select("td.total_MB").text(total_MB);
```
this update should happen after download, not in draw_geom

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -228,6 +228,9 @@ var animint = function (to_select, json_file) {
     g_info.tr.append("td").attr("class", "downloaded").text(0);
     g_info.tr.append("td").text(g_info.total);
+    g_info.tr.append("td").attr("class", "total_MB").text("0.00");
+    g_info.tr.append("td").attr("class", "mean_MB").text("0.00");
```
in these lines you should save the td so you can update it later withoug having to do select()
```suggestion
var td_mean_MB = g_info.tr.append("td").attr("class", "mean_MB");
```

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
Calculate chunk sizes in R, display in download table

- Calculate file sizes using file.size() in saveChunks()
- Store chunk_info (bytes, rows) in plot.json
- Update table cells after download (not in draw_geom)
- Remove 36-line JavaScript calculation loop

ANAMASGARD committed:
Bump version to 2025.10.31 and add NEWS entry for PR#272

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
Use spaces in column headers: 'total MB' and 'mean MB'

ANAMASGARD committed:
Add download size info to status table (#249)

- Remove 'selected chunk', 'status', and 'mean MB' columns
- Add combined 'files', 'MB', and 'rows' columns (downloaded/total format)
- Right-justify all numeric columns
- Fix common chunk tracking: count in both downloaded and total counts
- Track size (bytes/rows) for common chunks in R code
- Filter chunk_info to only include each geom's own chunks
- Count only .tsv files for accurate totals in JavaScript
- Update tests to verify new column structure

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
Fix MB calculation inconsistency in download table

Use 1048576 (1024*1024) consistently for MB conversion instead of
mixing 1e6 (1000000) and 1048576.

This ensures consistent size reporting for both common chunks and
regular chunks in the download status table.

ANAMASGARD committed:
Update download table tests for new files column

Switch test-download-status-table.R to validate rendered HTML instead of
the JS source. Verify headers and right-justified numeric columns via XPath.

Update test-renderer1-variable-value.R to read the combined
column ("downloaded / total") so the chunk count checks work again.

Remove the extra blank line in NEWS.md.

tdhock commented:
please click Resolve so I can see which previous comments you have already addressed

ANAMASGARD commented:
Sir @tdhock Here's the updated screenshot with the new download status table format:

<img width="3072" height="1920" alt="Screenshot From 2025-11-16 21-28-41" src="https://github.com/user-attachments/assets/b9424179-fe42-4df8-8010-aa50d9fba69a" />

**Changes implemented:**
- ✅ Removed "selected chunk" and "status" columns
- ✅ Removed "mean MB" column  
- ✅ Combined columns into "files", "MB", and "rows" format (downloaded/total)
- ✅ Right-justified all numeric columns
- ✅ Common chunk now counted in BOTH downloaded and total for all metrics

All values update correctly as data downloads, and common chunk is properly counted in both downloaded and total counts.

Please give your feedback and suggest what changes I need to make so that the failing test can be passed on the GitHub , Thank You !

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD commented:
Updated screenshot with WorldBank-paper.R example showing new download status table format:

<img width="3072" height="1920" alt="Screenshot From 2025-11-28 21-40-21" src="https://github.com/user-attachments/assets/900bba29-b929-4a69-9fa8-98da08c5acce" />


 - "files", "MB", and "rows" columns in "downloaded/total" format
 - Removed "selected chunk", "status", and "mean MB" columns
 - Right-justified numeric columns
 - Common chunk counted in both downloaded and total counts

tdhock commented:
that new screenshot looks good, thanks!
please fix test failures.

ANAMASGARD committed:
fix download status table: add id attribute and initialize display values

tdhock commented:
this needs to be fixed
```
── Failure ('test-renderer3-knit-print.R:67:3'): id property is unique over entire page ──
Expected `length(not.unique)` to equal 0.
Differences:
1/1 mismatches
[1] 1 - 0 == 1
```

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

tdhock commented on tests/testthat/test-download-status-table.R:
```diff
@@ -0,0 +1,28 @@
+  expect_true("rows" %in% header_text)
+  expect_false("status" %in% header_text)
+  expect_false("mean MB" %in% header_text)
+  expect_false("selected chunk" %in% header_text)
```
please remove

ANAMASGARD committed:
download status table: clean up tests and whitespace only

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
fix duplicate ID issue for knit-print compatibility

- Change download_status table ID from fixed 'download_status' to
  viz_id + '_download_status' for unique IDs when multiple animints
  are on the same page
- Update test XPath selectors to use contains() for flexibility

tdhock commented on NEWS.md:
```diff
@@ -10,6 +10,10 @@
 
+# Changes in version 2025.10.31 (PR#272)
+
+- Download status table now displays three new columns: `total_MB` (total disk space used by all downloaded chunks), `mean_MB` (average chunk size), and `rows` (total number of data rows). Chunk sizes are calculated in R using `file.size()` and exported via plot.json for efficient display updates after each chunk download.
```
please update

ANAMASGARD committed:
Update version to 2025.12.4 for PR#272

tdhock commented:
using this branch I see
<img width="492" height="388" alt="image" src="https://github.com/user-attachments/assets/15a967a4-acaa-44dd-9e6c-a48cbd798d7b" />

* large number of rows would benefit from having commas, like 4,321 instead of 4321 -- can you please fix?
* sometimes 0.00 is displayed for MB which could be confusing. I'm not sure what the best fix for this would be. Either change to 0 (without .00) or change column heading to "disk space" and use KB or MB in each row (depending on how big they are).

also please change the computation or column name "MB" to be consistent with "man du" which says
```
       The SIZE argument is an integer and optional unit (example: 10K is
       10*1024).  Units are K,M,G,T,P,E,Z,Y,R,Q (powers of 1024) or
       KB,MB,... (powers of 1000).  Binary prefixes can be used, too:
       KiB=K, MiB=M, and so on.
```

in the example above we see 15.62 MB for geom2.
but looking at the file sizes in R we see:
```r
> sum(size.dt[grep("geom2",name), R_bytes/1000/1000])
[1] 16.38009
> sum(size.dt[grep("geom2",name), R_bytes/1024/1024])
[1] 15.62127
```
15.62 is the number you get when you divide by 1024, which should be MiB (not MB as currently written in column header) so please 

* change MB to MiB in header, or
* change divisor from 1024 to 1000

thanks!

ANAMASGARD committed:
improve download status table: commas for rows, KiB/MiB for disk sizes

- Add commas to large row numbers (e.g., 4,321 instead of 4321)
- Show disk sizes in KiB or MiB based on size (binary 1024 divisor)
- Rename column from 'MB' to 'disk' since values now include units
- Update NEWS.md entry to reflect current features

ANAMASGARD committed:
Trigger CI re-run

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -993,8 +1059,20 @@ var animint = function (to_select, json_file) {
-        g_info.tr.select("td.downloaded").text(d3.keys(g_info.data).length);
         g_info.download_status[tsv_name] = "saved";
+        // Update size information after download
+        if(g_info.chunk_info && g_info.chunk_info[tsv_name]){
```
this if block is repeated above for common_tsv.
please refactor by adding a function to avoid this repetition.

tdhock reviewed (changes requested):
can you also please post a screenshot of the new display?

and please add a test for the content of the output? (current test is only for header names and alignment)

ANAMASGARD committed:
refactor download status table: add helper functions and content test

- Add updateDownloadStatus() function to eliminate code duplication
- Add formatWithCommas() for row numbers (4,321 instead of 4321)
- Add formatBytes() for KiB/MiB display using 1024 divisor
- Add test to verify table content format (files/disk/rows)

ANAMASGARD committed:
fix regex patterns in download status table test

Replace \d with [0-9] in regex patterns - \d is not recognized in R's base regex engine

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -225,10 +269,36 @@ var animint = function (to_select, json_file) {
+      // chunk_info includes the common chunk, so total_possible_chunks should include it
+      g_info.total_possible_chunks = tsv_count;
+    }
+
```
here and in test_that blocks

I've told you this before. can you please try to remember to do that next time, before asking for review?

tdhock commented:
the updated screenshot looks better, thanks!
it would be easier to read if there were a grey or black line to separate each row/column in the table. Can you add that please?

ANAMASGARD committed:
remove empty lines and add table borders

ANAMASGARD commented:
Yes Sir @tdhock  I will definitely create a function and post the updated screenshot.
Actually my laptop screen is broken,  ASAP it is fixed I will make the changes.

ANAMASGARD committed:
added applyCellStyles helper to avoid style repetition

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

ANAMASGARD committed:
fix legend naming regression and improve download status table test

ANAMASGARD committed:
Merge branch 'master' into add-download-size-info

tdhock commented on tests/testthat/test-download-status-table.R:
```diff
@@ -0,0 +1,42 @@
+  # Pattern allows digits with optional commas: 150 or 10,066
+  rows_pattern <- "^[0-9,]+ / [0-9,]+$"
+  rows_cells <- grep(rows_pattern, cell_text, value = TRUE)
+  expect_true(length(rows_cells) > 0, "Should have rows column with numeric format")
```
??

tdhock replied:
never mind, this diff is old

tdhock reviewed (approved):
ok
